### ISSUE TITLE
Export enhancements

### ISSUE THREAD

**Issue by bonny** [bonny/WordPress-Simple-History#170]
The export function could be a bit better:

- More export formats, for example formats that can include the diff of post (HTML? JSON?).
- Handle large logs without memory/timeouts. Create file using multipl AJAX calls and store in folder like `wp_privacy_generate_personal_data_export_file()` does.
- Make export function more visible by adding a link or similar to the main log file.
- Add option to export "what you see" while browsing the log, i.e. export the current "view".
- Maybe use https://actionscheduler.org/ for background processing.
- Add option to include CSV column headers, so users know what each column is

**Comment by AlecKinnear** [comment]
Hi Pär,

We love Simple History, and in large part for its simplicity. Simple History just works. Any of our clients who would like to track user activity on their membership sites, automatically get Simple History.

While we love the simplicity, we'd very much like to see an export button on the search results page. Here's how we imagine this could be done while staying true to the simplicity of Simple History.

Currently the two options for export are JSON or CSV. We are only after CSV here. To keep the interface simple you could add a setting for "Preferred Export Format" so that the button that appears on the search results page would just export in either CSV or JSON. If you add other export formats later, they could also be included in this preference.

Here's a version with the export button at the bottom:

![simple-history-interface-export-bottom](https://user-images.githubusercontent.com/107859958/218499189-beed2326-8d50-43e9-9075-d449726f21df.png)

Here's a version with the export button at the top:

![simple-history-interface-export-2a](https://user-images.githubusercontent.com/107859958/218499289-cf34171e-0eed-43f7-bcd8-44acbc1734dc.png)

Performance improvements on granular export are less of an issue than with the current global export. The big issue is that without granular export, there's no efficient way of getting data out of Simple History.

We have the skills to inject search results export ourselves into Simple History but would much prefer everyone enjoy the benefits. We feel strongly enough about this to contribute $150 to Simple History if search result export is coded this month (yes, there's time pressure, I'm sorry). Alas, if you were unable to make this improvement now, we'd have to do it ourselves.

Let us know if you are able to make this improvement, Pär, or if you have other ideas about it.

Once export is in, we'll be using the ability to register loggers to improve search and filtering for our events (FV Player). Thanks including the calls and filters which allow Simple History to play well with others.

**Comment by bonny** [comment]
@AlecKinnear Glad to hear that you like the plugin! 

I agree that the export needs some more work. At the moment development of the plugin is a bit slow due to lack of time, but hopefully I will have time to work on it more soon again. 

Don't know if it's good enough for you but you could probably pretty easy create a "box" in the sidebar and add a link/button there that goes to the export page. It won't include the filters but it may be better than nothing. :/

**Comment by AlecKinnear** [comment]
@bonny Thanks for the quick response. I didn't write back right away as I found your answer so frustrating. We need filters, we need selective export. I gave exact guidelines on how to build export functionality to be genuinely useful in a few hours. On top of that we offered to pay for the work (albeit $150). As a demonstration of good faith, I started to sponsor your work on Github.

We do not want to do this enhancement ourselves just for ourselves. We'd like you to do the selective CSV export enhancement **now** and get a contribution and have _everyone_ enjoy the improved export functionality.

And at some point in a mythical future you can add html, json, xml whatever catches your fancy. You've set the bar too high and are letting everyone down by doing so. Make it easy. Just get selective export for CSV done, so the vast majority of your users can export data in a useful format.

I'm very tired of open source developers complaining that no one contributes and then when people come forward to contribute, the open source developers complain they have no time and are bored with the project anyway. This is why we cannot enjoy high quality FOSS software. And instead have to overpay for over-marketed buggy commercial software run by ethically challenged MBA types or get-rich-quicksters. Yes, I'm frustrated Pär.

**Comment by bonny** [comment]
I understand that the export functionality is important to you so I understand the frustration. I also have reported issues in projects that received no replies or that gets closed without being fixed. It’s indeed very annoying.

Many other open source developers probably have the same problem that I have: a family and a full-time job, so it’s challenging to squeeze in more than an hour here and there to work on OSS projects.

**Comment by AlecKinnear** [comment]
@Pär I understand open source. I'm not asking you to work for free. I signed up to sponsor you on Github *and* promised $150/€150 (take your pick) to implement selective CSV import which is as described above is just a couple of hours of your time (half an hour to settle down and remember the code, forty minutes to write a solution, fifty minutes to test). Given that I'd probably sponsor you for a year if I was happy with what happened with this request, you are effectively refusing to improve export at all in Simple History for €220.

Of course the version you are talking about with multiple formats and some kind of special interface (which I deliberately did not ask for, just use the existing search function which prepares a selection already) would take about 25 hours to code and even then would be full of bugs and require continuous additions and fixes.

We implement feature requests all the time for our customers if they are willing to pay for it. If open-source developers won't accept sponsorship to improve their projects, the whole FOSS is doomed. We (we're FOSS developers as well) just can't seem to figure out a functional model at all. And it's a great pity.

Next question. If we go through the absolute misery of leaning Simple History's code well enough (we're half way there, your code is clean at least if the named variables are a little bit nested) to code the CSV export ourselves, will you accept the push request? I expect the answer will be no, as it's a lot of trouble to manage external code contributions. Which is exactly why I offered to donate to the project in the first place.

What a great contribution (Simple History) to WordPress just being left to atrophy and rot. All hail FOSS! Instead we will all have to deal with absolute crap adware tracking solutions full of admin notice interruptions, ads and annual renewal licenses. What's even better about WordPress these days are the plugin authors who claim their plugins should stop working after the license expires.

Rant over – it's not your fault that FOSS doesn't work and that WordPress has become a viper's den of spammy marketing wares instead of collective code which we developers all carefully cultivate in cheerful and considerate co-operataion.

**Comment by AlecKinnear** [comment]
@bonny I checked your history of merging pulls from Simple History. It's very good. Hence, as we can't persuade you to improve CSV export in Simple History for love or money, we'll make some improvements and send you the pull request. I'd still rather send you the sponsorship money and that you made the improvements. Let me know.

**Comment by AlecKinnear** [comment]
Hi Pär,

We just need simple CSV export and that would help most users. I've withdrawn sponsorship as there's not enough movement on this issue and I've sponsored Simple History for half a year. We have had good use of Simple History, so it's no hard feelings. But if open source authors would like to maintain sponsors, it would help to heed their feature requests. If simple CSV export appears some time soon (I see you've made some progress on some internal classes), I'll be happy to renew our sponsorship.

**Comment by AlecKinnear** [comment]
Can't get this done for love or money. It's not a matter of having a family, I was willing to sponsor the feature, within an achievable format (not the max version which would be a couple of weeks coding rather than a few hours). Argh.

Open source authors complain there's no financial support, but even when there is it's impossible to get improvements in a timely way. Thanks for Simple History all the same!

**Comment by bonny** [comment]
In [Simple History 5.4](https://simple-history.com/2024/simple-history-5-4-0/), together with the [Premium add-on](https://simple-history.com/add-ons/premium/), it is now possible to do CSV (and JSON + HTML) exports. This is thanks to the rewrite in React 💪 .

The export feature is the first feature/action to use the new [Event Actions Toolbar](https://simple-history.com/2024/simple-history-5-4-0/#event-actions-toolbar) (or should it be called control bar? 🤔)

![Image](https://github.com/user-attachments/assets/a7dd7cf1-04a7-40d8-a1f9-ec662310b6db)

**Comment by AlecKinnear** [comment]
Hi @bonny 

This is exciting news.

Except that you are not asking even for a Premium add-on but for another subscription and an expensive one. We use Simple History on about twenty sites. To do so should now cost us $600/year?

What happened to the good old WordPress of developers helping other developers? From free to extremely expensive and limited subscription.

We would have been up for a one-time payment or to contribute to Simple History for this feature. But subscription, no way.

**Comment by bonny** [comment]
Thanks for the feedback @AlecKinnear. Give it a try - if you don't fancy it I'll just refund you the money.

Also, I do agree that $600 is much. I will look over the pricing and make sure there are options for users with many sites.

**Comment by AlecKinnear** [comment]
Hi Per,

Thanks for your note. I had already bought a five-site license with the New Year's discount. I had promised you support for the plugin* and honour my word.

What bothers me about the marketing of Simple History is that it stinks of the WooCommerce *banditisme* which then spread out across the WordPress ecosphere (including WP Search for instance, although Jon Christopher has since sold off WP Search to Syed Balkhi of all people). The way it works is that follow-up payments are the same price as the initial purchase. Since the subscription payments are so ridiculously high, it's bait and switch. A discount on the first year, and a forced/hidden subscription at double the price.

This is the opposite of how historically software pricing worked. The first payment should be higher and subsequent renewals are at a discount. Bringing on a new users is the biggest load for us developers (initial support tickets), renewals are easy. This model shows a lot more respect for one's users and makes the long term engagement more reasonable. I'm not sure it should always be subscription. Some software should be sold outright. For those of us working inside a constantly changing ecosphere, annual payments make more sense. We spend more time keeping our software working properly in the face of four new OS/year (macOS, Windows, iOS, Android) plus all the changes to internet protocols (this month iCloud Private Relay is creating all kinds of trouble for us with secure video playback for instance).

Any way, the point is that just because the bad actors (Matt Mullenweg, Yoast, Syed Balkhi) behave badly is not a good reason for us (people who really believe in Open Source and are not secretly trying to privatise the web**) for us to adopt the bad norms which they try to force on us. You will argue, it works, it pays to behave like this. I would argue that:

1. it's still wrong. 
2. it doesn't work.

With WooCommerce, we've paid for many add-ons but these days we avoid installing it on client sites unless really needed where we use nulled versions most of the time. WooCommerce is stolen software, more or less without support, with ludicrous licensing terms. No, Automattic cannot extract $2000/year per site *ad infinitum* from either us or our clients. When we do try to use WooCommerce support (every once in a while, though to little avail), we do pay for that add-on that year.

In the case of Simple History, instead of giving you $100/year, we've paid once. I don't know when we'll pay again. Simple History is now a commercial plugin, which makes the relationship different. I would have been happy to give you $10/month in perpetuity knowing that Simple History is helping other developers. It's difficult though as most users won't pay unless the developer demands payment.

Again, I have no issue with you charging for Simple History. You deserve it, what I dislike though is this bait and switch of trying to lure users into expensive subscriptions.

With our own software, renewals used to be at 1/3 of first purchase (we wanted to sell one-time licenses but it's impossible). We've moved to continuous renewals at 1/2 price. Interrupted renewal is at 3/4 of full price. We wanted to motivate our users to continuously pay for support and updates, rather than every two or three years. It's mostly worked. I've been hardline with users when they miss the renewal deadline and submit a support ticket asking for the discount, as support is most software company's single biggest cost. But it did teach me to remind users not once but twice about the upcoming renewal deadline.

Great that you have built some kind of commercial structure to support ongoing improvement of Simple History. My thoughts above are not binding. These are just the thoughts of developers who have been using Simple History a long time, contributing to Open Source and WordPress from the beginning. It's difficult to skate between the extremes of FOSS (no payment for anything) and the barracuda-like exploitation of users (Adobe, Microsoft, Oracle, WooCommerce, Yoast, Balkhi), executed more and more often under a mantle of Open Source.

Keep fighting the good fight!

Making the web work for you, Alec

(*) which I would have been happier to provide as a regular monthly donation for faster action on the one feature we wanted, which was simple CSV export

(**) Yes, I include Matt in the group of baddies, as he's the one who pioneered stealing other people's work and profiting off of it and encouraged special treatment in the plugin directory for his own companies, his buddies and WordCamp sponsors. At least until it didn't suit him any more at all, at which point he moved to straight shakedowns (though WP Engine are anything but saints themselves, selling underpowered VPS at ultra-premium rates and trying to privatise WordPress themselves, a pox on all their houses).
